File: packages/python/src/openskp/export/stl.py
"""STL (Standard Triangle Language) export module for OpenSKP.

Exports a baked :class:`~openskp.scene.Scene` to STL format (ASCII or Binary)
for 3D printing and CAD interchange.
"""

from __future__ import annotations

import math
import pathlib
import struct
from typing import Union

from ..scene import Scene


def _calculate_normal(
    v0: tuple[float, float, float],
    v1: tuple[float, float, float],
    v2: tuple[float, float, float],
) -> tuple[float, float, float]:
    """Calculate normalized normal vector from 3 vertices."""
    e1x, e1y, e1z = v1[0] - v0[0], v1[1] - v0[1], v1[2] - v0[2]
    e2x, e2y, e2z = v2[0] - v0[0], v2[1] - v0[1], v2[2] - v0[2]

    nx = e1y * e2z - e1z * e2y
    ny = e1z * e2x - e1x * e2z
    nz = e1x * e2y - e1y * e2x

    length = math.sqrt(nx * nx + ny * ny + nz * nz)
    if length > 1e-12:
        return (nx / length, ny / length, nz / length)
    return (0.0, 0.0, 0.0)


def to_stl_ascii(scene: Scene, scale: float = 1.0) -> str:
    """Serialize a baked scene to ASCII STL text format.

    Args:
        scene: The baked scene returned by :meth:`SkpFile.build_scene`.
        scale: Scale factor for vertex coordinates (e.g. 1.0 for metres,
            1000.0 to convert metres to millimetres for 3D slicers).

    Returns:
        The formatted ASCII STL text string.
    """
    lines: list[str] = ["solid OpenSKP_Model"]

    for prim in scene.glb_primitives:
        tri_count = len(prim.indices) // 3
        for i in range(tri_count):
            i0 = prim.indices[i * 3]
            i1 = prim.indices[i * 3 + 1]
            i2 = prim.indices[i * 3 + 2]

            v0 = (
                prim.positions[i0 * 3] * scale,
                prim.positions[i0 * 3 + 1] * scale,
                prim.positions[i0 * 3 + 2] * scale,
            )
            v1 = (
                prim.positions[i1 * 3] * scale,
                prim.positions[i1 * 3 + 1] * scale,
                prim.positions[i1 * 3 + 2] * scale,
            )
            v2 = (
                prim.positions[i2 * 3] * scale,
                prim.positions[i2 * 3 + 1] * scale,
                prim.positions[i2 * 3 + 2] * scale,
            )

            nx, ny, nz = _calculate_normal(v0, v1, v2)

            lines.append(f"  facet normal {nx:.6f} {ny:.6f} {nz:.6f}")
            lines.append("    outer loop")
            lines.append(f"      vertex {v0[0]:.6f} {v0[1]:.6f} {v0[2]:.6f}")
            lines.append(f"      vertex {v1[0]:.6f} {v1[1]:.6f} {v1[2]:.6f}")
            lines.append(f"      vertex {v2[0]:.6f} {v2[1]:.6f} {v2[2]:.6f}")
            lines.append("    endloop")
            lines.append("  endfacet")

    lines.append("endsolid OpenSKP_Model\n")
    return "\n".join(lines)


def to_stl_binary(scene: Scene, scale: float = 1.0) -> bytes:
    """Serialize a baked scene to Binary STL format.

    Args:
        scene: The baked scene returned by :meth:`SkpFile.build_scene`.
        scale: Scale factor for vertex coordinates (e.g. 1.0 for metres,
            1000.0 to convert metres to millimetres for 3D slicers).

    Returns:
        The packed Little-Endian Binary STL byte array.
    """
    total_triangles = sum(len(p.indices) // 3 for p in scene.glb_primitives)

    # 80-byte header
    header = b"# OpenSKP Binary STL Export"
    header = header.ljust(80, b"\x00")

    chunks: list[bytes] = [header, struct.pack("<I", total_triangles)]

    for prim in scene.glb_primitives:
        tri_count = len(prim.indices) // 3
        for i in range(tri_count):
            i0 = prim.indices[i * 3]
            i1 = prim.indices[i * 3 + 1]
            i2 = prim.indices[i * 3 + 2]

            v0 = (
                prim.positions[i0 * 3] * scale,
                prim.positions[i0 * 3 + 1] * scale,
                prim.positions[i0 * 3 + 2] * scale,
            )
            v1 = (
                prim.positions[i1 * 3] * scale,
                prim.positions[i1 * 3 + 1] * scale,
                prim.positions[i1 * 3 + 2] * scale,
            )
            v2 = (
                prim.positions[i2 * 3] * scale,
                prim.positions[i2 * 3 + 1] * scale,
                prim.positions[i2 * 3 + 2] * scale,
            )

            nx, ny, nz = _calculate_normal(v0, v1, v2)

            # Pack: 3 floats normal + 3x3 floats vertices + 1 uint16 attribute count
            tri_data = struct.pack(
                "<12fH",
                nx,
                ny,
                nz,
                v0[0],
                v0[1],
                v0[2],
                v1[0],
                v1[1],
                v1[2],
                v2[0],
                v2[1],
                v2[2],
                0,
            )
            chunks.append(tri_data)

    return b"".join(chunks)


def export(
    scene: Scene,
    output_path: Union[str, pathlib.Path],
    binary: bool = False,
    scale: float = 1.0,
) -> None:
    """Export a baked scene to an STL file.

    Args:
        scene: The baked scene returned by :meth:`SkpFile.build_scene`.
        output_path: Destination file path (.stl).
        binary: If True, writes binary STL. Otherwise writes ASCII STL.
        scale: Scale factor for vertex coordinates (e.g. 1000.0 for mm).
    """
    if scene is None:
        raise ValueError("scene cannot be None")

    out = pathlib.Path(output_path)
    out.parent.mkdir(parents=True, exist_ok=True)

    if binary:
        data = to_stl_binary(scene, scale=scale)
        with open(out, "wb") as fp:
            fp.write(data)
    else:
        text = to_stl_ascii(scene, scale=scale)
        with open(out, "w", encoding="utf-8") as fp:
            fp.write(text)


__all__ = ["to_stl_ascii", "to_stl_binary", "export"]

File: packages/typescript/src/stl.ts
import { SkpScene } from './model';

declare const process: any;
declare const require: any;

function calculateNormal(
  v0: [number, number, number],
  v1: [number, number, number],
  v2: [number, number, number]
): [number, number, number] {
  const e1x = v1[0] - v0[0];
  const e1y = v1[1] - v0[1];
  const e1z = v1[2] - v0[2];

  const e2x = v2[0] - v0[0];
  const e2y = v2[1] - v0[1];
  const e2z = v2[2] - v0[2];

  const nx = e1y * e2z - e1z * e2y;
  const ny = e1z * e2x - e1x * e2z;
  const nz = e1x * e2y - e1y * e2x;

  const len = Math.sqrt(nx * nx + ny * ny + nz * nz);
  if (len > 1e-12) {
    return [nx / len, ny / len, nz / len];
  }
  return [0, 0, 0];
}

/**
 * Serialize a baked SkpScene into ASCII STL text format.
 *
 * @param scene The result of SkpFile.buildScene()
 * @param scale Optional scale factor (e.g. 1000.0 for mm)
 * @returns The formatted ASCII STL string.
 */
export function toSTLAscii(scene: SkpScene, scale = 1.0): string {
  if (!scene || !scene.glbPrimitives) {
    throw new Error('toSTLAscii requires a valid SkpScene instance');
  }

  const lines: string[] = ['solid OpenSKP_Model'];

  for (const prim of scene.glbPrimitives) {
    const triCount = Math.floor(prim.indices.length / 3);
    for (let i = 0; i < triCount; i++) {
      const i0 = prim.indices[i * 3];
      const i1 = prim.indices[i * 3 + 1];
      const i2 = prim.indices[i * 3 + 2];

      const v0: [number, number, number] = [
        prim.positions[i0 * 3] * scale,
        prim.positions[i0 * 3 + 1] * scale,
        prim.positions[i0 * 3 + 2] * scale,
      ];
      const v1: [number, number, number] = [
        prim.positions[i1 * 3] * scale,
        prim.positions[i1 * 3 + 1] * scale,
        prim.positions[i1 * 3 + 2] * scale,
      ];
      const v2: [number, number, number] = [
        prim.positions[i2 * 3] * scale,
        prim.positions[i2 * 3 + 1] * scale,
        prim.positions[i2 * 3 + 2] * scale,
      ];

      const [nx, ny, nz] = calculateNormal(v0, v1, v2);

      lines.push(
        `  facet normal ${nx.toFixed(6)} ${ny.toFixed(6)} ${nz.toFixed(6)}`
      );
      lines.push('    outer loop');
      lines.push(
        `      vertex ${v0[0].toFixed(6)} ${v0[1].toFixed(6)} ${v0[2].toFixed(6)}`
      );
      lines.push(
        `      vertex ${v1[0].toFixed(6)} ${v1[1].toFixed(6)} ${v1[2].toFixed(6)}`
      );
      lines.push(
        `      vertex ${v2[0].toFixed(6)} ${v2[1].toFixed(6)} ${v2[2].toFixed(6)}`
      );
      lines.push('    endloop');
      lines.push('  endfacet');
    }
  }

  lines.push('endsolid OpenSKP_Model\n');
  return lines.join('\n');
}

/**
 * Serialize a baked SkpScene into Little-Endian Binary STL format.
 *
 * @param scene The result of SkpFile.buildScene()
 * @param scale Optional scale factor (e.g. 1000.0 for mm)
 * @returns Packed Little-Endian Uint8Array.
 */
export function toSTLBinary(scene: SkpScene, scale = 1.0): Uint8Array {
  if (!scene || !scene.glbPrimitives) {
    throw new Error('toSTLBinary requires a valid SkpScene instance');
  }

  let totalTriangles = 0;
  for (const prim of scene.glbPrimitives) {
    totalTriangles += Math.floor(prim.indices.length / 3);
  }

  const bufferSize = 80 + 4 + totalTriangles * 50;
  const buffer = new ArrayBuffer(bufferSize);
  const view = new DataView(buffer);

  // Write 80-byte header
  const headerText = '# OpenSKP Binary STL Export';
  for (let i = 0; i < 80; i++) {
    const charCode = i < headerText.length ? headerText.charCodeAt(i) : 0;
    view.setUint8(i, charCode);
  }

  // Write triangle count
  view.setUint32(80, totalTriangles, true);

  let offset = 84;
  for (const prim of scene.glbPrimitives) {
    const triCount = Math.floor(prim.indices.length / 3);
    for (let i = 0; i < triCount; i++) {
      const i0 = prim.indices[i * 3];
      const i1 = prim.indices[i * 3 + 1];
      const i2 = prim.indices[i * 3 + 2];

      const v0: [number, number, number] = [
        prim.positions[i0 * 3] * scale,
        prim.positions[i0 * 3 + 1] * scale,
        prim.positions[i0 * 3 + 2] * scale,
      ];
      const v1: [number, number, number] = [
        prim.positions[i1 * 3] * scale,
        prim.positions[i1 * 3 + 1] * scale,
        prim.positions[i1 * 3 + 2] * scale,
      ];
      const v2: [number, number, number] = [
        prim.positions[i2 * 3] * scale,
        prim.positions[i2 * 3 + 1] * scale,
        prim.positions[i2 * 3 + 2] * scale,
      ];

      const [nx, ny, nz] = calculateNormal(v0, v1, v2);

      // Normal (3x Float32)
      view.setFloat32(offset, nx, true);
      view.setFloat32(offset + 4, ny, true);
      view.setFloat32(offset + 8, nz, true);

      // Vertex 0 (3x Float32)
      view.setFloat32(offset + 12, v0[0], true);
      view.setFloat32(offset + 16, v0[1], true);
      view.setFloat32(offset + 20, v0[2], true);

      // Vertex 1 (3x Float32)
      view.setFloat32(offset + 24, v1[0], true);
      view.setFloat32(offset + 28, v1[1], true);
      view.setFloat32(offset + 32, v1[2], true);

      // Vertex 2 (3x Float32)
      view.setFloat32(offset + 36, v2[0], true);
      view.setFloat32(offset + 40, v2[1], true);
      view.setFloat32(offset + 44, v2[2], true);

      // Attribute byte count (Uint16)
      view.setUint16(offset + 48, 0, true);

      offset += 50;
    }
  }

  return new Uint8Array(buffer);
}

/**
 * Export a baked SkpScene directly to an STL file.
 * Node.js environment only.
 *
 * @param scene The result of SkpFile.buildScene()
 * @param outputPath Destination file path (.stl)
 * @param options Export options (binary format flag, scale multiplier)
 */
export function exportSTL(
  scene: SkpScene,
  outputPath: string,
  options?: { binary?: boolean; scale?: number }
): void {
  if (typeof process !== 'undefined' && process.versions && process.versions.node) {
    const fs = require('fs');
    const path = require('path');
    const binary = options?.binary ?? false;
    const scale = options?.scale ?? 1.0;

    const dir = path.dirname(outputPath);
    if (dir && !fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    if (binary) {
      const data = toSTLBinary(scene, scale);
      fs.writeFileSync(outputPath, data);
    } else {
      const text = toSTLAscii(scene, scale);
      fs.writeFileSync(outputPath, text, 'utf-8');
    }
  } else {
    throw new Error('exportSTL file writing is only supported in Node.js environment');
  }
}
